README.md
# HEncoder-HDecoder [![Build Status](https://travis-ci.org/dragonhades/HEncoder-HDecoder.svg?branch=master)](https://travis-ci.org/dragonhades/HEncoder-HDecoder)  
Encoder and Decoder based on the user input password to encode the given ASCII text file using Huffman tree.
  
[中文说明](https://github.com/dragonhades/HEncoder-HDecoder/blob/master/README-zh.md).  
  
Install:
---------
  Please download all files in order to compile successfully. Makefile will generate two executables, "HDecode" and "HEncode". Compiler used is g++-5 -std=c++14

Usage:  
--------
  * 'HEncode' takes a filename as commandline arguement. It will ask the user to give input as a password, and will output the encoded binary file to stdout.  
  * 'HDecode' takes a filename as commandline arguement. It will ask the user to give input as a password, and ask again to prevent possible typos. Then it will output the decoded text to stdout.  
  * 'runTest.sh' is a bash script yo help testing. It takes a filename as commandline argument, and use 'diff' command as result of output.  

Demo:
-----
```bash
$ ./HEncode toEncode.txt > encode.out  
Enter password:  
pw42  
Please enter again:  
pw42  
$ ./HDecode encode.out > myfile.txt  
Enter password:  
WARNING: Wrong password will get you unreadable messy code.  
pw42  
$ diff toEncode.txt myfile.txt  
$  
```
  
Note:
-----
 * The input file should only contain ASCII charators.  
 * Only when encoding password and decoding password match so the correct original file will be outputed.  
   If the password given in HDecode is wrong, unreadable messy code will be given.  

Credits: 
--------
http://stackoverflow.com/questions/6899025/hide-user-input-on-password-prompt   for hiding user password input from screen  
  
Contact: 
--------
Any bug reports or concerns please contact jimmyshanofficial@gmail.com 


news.txt
Donald Trump has said the United States will "solve" the nuclear threat from North Korea, with or without China's help.

"If China is not going to solve North Korea, we will. That is all I am telling you," he said in an interview with UK paper Financial Times.

Pressed on whether he thought he could succeed alone, he replied: "Totally."

Mr Trump was speaking ahead of a scheduled visit from Chinese President Xi Jinping this week.

"China has great influence over North Korea. And China will either decide to help us with North Korea, or they won't. And if they do that will be very good for China, and if they don't it won't be good for anyone," Mr Trump told the FT.

Asked if he meant "one-on-one" unilateral action, Mr Trump said: "I don't have to say any more. Totally."

He did not give any further details on what action he would take.

Mr Trump's brief comments, published just days before the key meeting with Mr Xi at Mar-a-Lago on Thursday, are the latest in a series of warnings over North Korea's nuclear development.

There are fears that Pyongyang could eventually develop the ability to launch long-range nuclear missiles capable of striking the mainland United States.


test/runTest.sh
#!/bin/bash
suite=$(mktemp)
ls | egrep "*\.txt$" > $suite
stat=1
for input in $(cat $suite); do
  in1=$(mktemp)
  in2=$(mktemp)
  out=$(mktemp)
  file=$(mktemp)
  echo pwd >> ${in1}
  echo pwd >> ${in1}
  echo pwd > ${in2}
  cat ${in1} | ../HEncode ${input} > ${out} 2> /dev/null
  cat ${in2}| ../HDecode ${out} > ${file} 2> /dev/null
  diff ${input} ${file} > /dev/null
  if [ $? -eq 1 ]; then 
    echo "Test ${input} failed!"
    rm ${in1} ${in2} ${out} ${file}
    stat=0
  else 
    echo "Test passed!"
    rm ${in1} ${in2} ${out} ${file}
  fi
done
rm ${suite}
if [ ${stat} -eq 0 ]; then
  exit 1
fi


tools.h
#ifndef HEADER
#define HEADER

// int defined in tool.cc, see more details there
// Used to set the acc for huffman coding loop sequence
// Default to 128
extern unsigned int ACCUMULATOR;

struct Ascii {
  std::vector<int> freq;
  Ascii();
  void update(const int index);
  void print();
  void hufffreq();
  void clear();
};

class Node {
  int c;
  int weight;
public:
  std::shared_ptr<Node> left=nullptr;
  std::shared_ptr<Node> right=nullptr;
  Node(const int c, const int weight):c{c},weight{weight}{}
  std::string out(const int c, const std::string &s);
  int get_weight() const ;
  int get_intChar() const ;
};

class Tree {
  std::vector<Node>list;
public:
  Tree(const Ascii &a);
  Node front() const ;
  std::vector<Node>::iterator begin() ;
  std::vector<Node>::iterator end() ;
  void merge_least();
  void huffcode();
  void print();
};

void setSeed(std::string &s);

int depth(const Node &n, int inp);

void wpl(Tree &t, Ascii &a);

#endif


tools.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <memory>
#include <algorithm>
#include <random>
using namespace std;

#include "tools.h"

// Global int defined for HEncode.cc and HDecode.cc to adjust the acc for 
//  huffman loops. Default to 128
// Since the number of all outcome of random_shuffle is ACCUMULATOR! (factorial)
//  smaller the nummber, more unsafe it gets.
// But the huffman compression works better if the file charator count 
//  exceeds ACCUMULATOR by great amount.
// Hence, change this number accordingly to get the best compression results.
// WARNING: the program does not check the validity of this number.
//          Any unexpected results due to the change of this number is not responsible
//          by the program.
unsigned int ACCUMULATOR=128;

int Node::get_weight() const {
  return weight;
}

int Node::get_intChar() const {
  return c;
}

string Node::out(const int inp, const string &s){
  if(inp == c){
    return s;
  } else {
    if((!left)&&(!right)) throw 0;
    if(!left){
      return right->out(inp,s+"1");
    }
    try {
      return left->out(inp,s+"0");
    } catch(...){
      if(!right) throw 0;
      else return right->out(inp,s+"1");
    }
  }
}

Tree::Tree(const Ascii &a){
  for(register int i=0;i<128;i++){
    list.emplace_back(i,a.freq[i]);
  }
}

Node Tree::front() const {
  if(list.empty()) throw "List is empty";
  return list[0];
}

vector<Node>::iterator Tree::begin() {
  return list.begin();
}

vector<Node>::iterator Tree::end() {
  return list.end();
}

void Tree::huffcode(){
  ofstream of("huff.code");
  for(register int i=0;i<128;i++){
    string s = list[0].out(i,"");
    of<<s<<endl;
  }
}

void Tree::merge_least(){
  if(int(list.size())==1) return;
  int min = list[0].get_weight();
  auto least=list.begin();
  for(auto l=list.begin();l!=list.end();l++){
    if(l->get_weight()<min) {
      min = l->get_weight();
      least=l;
    }
  }
  Node temp = *least;
  list.erase(least);
  min = list[0].get_weight();
  auto second=list.begin();
  for(auto l=list.begin();l!=list.end();l++){
    if(l->get_weight()<min){
      min = l->get_weight();
      second = l;
    }
  }
  int left = temp.get_weight();
  int right = second->get_weight();
  Node n(-1,left+right);
  n.left = make_shared<Node>(temp);
  n.right = make_shared<Node>(*second);
  list.erase(second);
  list.emplace_back(n);
  return merge_least();
}

Ascii::Ascii(){
  for(register int i=0;i<128;i++){
    freq.emplace_back(0);
  }
}

void Ascii::update(const int index){
  freq[index]++;
}

void Ascii::print(){
  for(auto it:freq){
    cerr<<it<<endl;
  }
}

void Ascii::hufffreq(){
  ofstream of("huff.freq");
  for(auto it:freq){
    of<<it<<endl;
  }
}

void Ascii::clear(){
  freq.clear();
  for(register int i=0;i<128;i++){
    freq.emplace_back(0);
  }
}

void setSeed(string &s){
  seed_seq seed(s.begin(),s.end());
  default_random_engine generator(seed);
  int gen = generator();
  srand(gen);
}

int depth(const Node &n, int inp){
  if(inp == n.get_intChar()){
    return 0;
  } else {
    if((!n.left)&&(!n.right)) throw 0;
    if(!n.left){
      return 1+depth(*n.right,inp);
    }
    try {
      return 1+depth(*n.left,inp);
    } catch(...){
      if(!n.right) throw 0;
      else return 1+depth(*n.right,inp);
    }
  }
}

void wpl(Tree &t, Ascii &a){
  int sum=0;
  for(int i=0; i<128; i++){
    int freq = a.freq[i];
    if(freq==0) continue;
    int dep = depth(t.front(),i);
    sum += freq*dep;
  }
  ofstream of("huff.wpl");
  of<<sum<<endl;
}	


HDecode.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <limits>
#include <bitset>
#include <vector>
#include <stack>
#include <memory>
#include <algorithm>
#include <termios.h>
#include <unistd.h>
using namespace std;

#include "tools.h"

class Tree;
class Node;
struct Ascii;

// int defined in tool.cc, see more details there
// Used to set the acc for huffman coding loop sequence
// Default to 128
extern unsigned int ACCUMULATOR;

// open filename and read 1 byte every time, get bit and convert them into a string
string read_bits(const char* filename){
  string bitstring;
  ifstream file(filename, ios::binary);
  if(file.fail()) throw string("Cannot open ")+filename;
  char c;
  while(file.get(c)){
    for (int i = 7; i >= 0; i--) {
      if(int((c >> i) & 1)==1) bitstring += "1";
      else bitstring += "0";
    }
  }
  return bitstring;
}

// use Tree t, find and return the decoded message
char decode(Tree &t, vector<char> &input){
  int inp;
  auto ptr = make_shared<Node>(t.front());
  while(1){
    if(ptr->left==nullptr&&ptr->right==nullptr){
      return (char) ptr->get_intChar();
    }
    if(input.empty()) throw "EMPTY";
    inp = input[0];
    input.erase(input.begin());
    if(inp=='0'){
      ptr = ptr->left;
    } else {
      ptr = ptr->right;
    }
  }
  throw "EMPTY";
}

int main(int argc, char* argv[]){
  
  // check if have enough argument
  if(argc<2){
    cerr<<"Usage: ./HDecode Filename";
    return 0;
  }

  string str;
  try {
    str = read_bits(argv[1]);
  } catch(const string &s){
    cerr<<s<<endl;
    return 0;
  }
  
  // hide user entered password from screen
  //http://stackoverflow.com/questions/6899025/hide-user-input-on-password-prompt
  termios oldt;
  tcgetattr(STDIN_FILENO, &oldt);
  termios newt = oldt;
  newt.c_lflag &= ~ECHO;
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  
  cerr<<"Enter password:"<<endl;
  cerr<<"WARNING: Wrong password will get you unreadable messy code."<<endl;
  string enter;
  getline(cin, enter);
  setSeed(enter);

  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);

  char inp;
  Ascii ascii1;
  Ascii ascii2;
  vector<char> input;
  int acc = ACCUMULATOR;
  int count = 0;
  istringstream iss(str);
  while(iss>>inp){
    input.emplace_back(inp);
  }

  while(!input.empty()){
    
    Tree t(ascii1);
    
    if(count==0){
      // see HEncode.cc for details
      random_shuffle(t.begin(),t.end());
    }
    t.merge_least();
    
    if(count==1) acc = ACCUMULATOR;  // need a sequence of acc, acc, acc*2, acc*4, acc*8 ...
    
    for(register int i = 0; i<acc; i++){
      char c;
      try {
	c = decode(t,input); // may throw 
      } catch(const char *exp){ break; }
      unsigned int num = int(c);
      ascii2.update(num);
      cout<<c;
    }
    
    acc=acc*2;
    count++;
    ascii1=ascii2;
  }
}


HEncode.cc
#include <iostream>
#include <fstream>
#include <vector>
#include <memory>
#include <random>
#include <algorithm>
#include <cstdio>
#include <sstream>
#include <fstream>
#include <termios.h>
#include <unistd.h>
using namespace std;

#include "tools.h"

class Tree;
class Node;
struct Ascii;

// int defined in tool.cc, see more details there
// Used to set the acc for huffman coding loop sequence
// Default to 128
extern unsigned int ACCUMULATOR;

int char_to_int(const char &c){
  if(c=='0') return 0;
  else return 1;
}

void get_bit(int *arr, int &count, const char c){
  count++;
  arr[count-1] = char_to_int(c);
  if(count==8) count=0;
}

// using t as Huffman tree to encode input, return 0 & 1 in a string.
string encode(Tree &t,vector<int> &input){
  string s;
  for(auto num:input){
    s += t.front().out(num,"");
  }
  return s;
}

// bin_out outputs encoding into binary
// Challenges are that C++ (most languages) only can output 1 byte (1 char) at a time 
// Need to concatenate all bits into bytes intuitively
void bin_out(const string &s){
  istringstream iss(s);
  char c;
  int count=0;	// counting the bits
  int arr[8];	// use to contain one byte.
  
  try {
    while(1){
      for(register int i=0;i<8;i++){
	if(!(iss>>c)) throw 0;	// when input exhausted, need to "round up" a byte
	get_bit(arr, count, c);	//convert c into int and put in arr, count++
      }

      // if we have a full byte, we output it.
      putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5|arr[3]<<4|arr[4]<<3|arr[5]<<2|arr[6]<<1|arr[7]);
    
    }
  } catch(...){

    // input exhausted, round up a byte by filling 0 bit(s).

    switch(count){
      case 0:
	break;
      case 1:
	putchar(arr[0]<<7);
	break;
      case 2:
	putchar(arr[0]<<7|arr[1]<<6);
	break;
      case 3:
	putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5);
	break;
      case 4:
	putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5|arr[3]<<4);
	break;
      case 5:
	putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5|arr[3]<<4|arr[4]<<3);
	break;
      case 6:
	putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5|arr[3]<<4|arr[4]<<3|arr[5]<<2);
	break;
      case 7:
	putchar(arr[0]<<7|arr[1]<<6|arr[2]<<5|arr[3]<<4|arr[4]<<3|arr[5]<<2|arr[6]<<1);
	break;
    }
  }
}

int main(int argc, char* argv[]){
  
  // check if have enough argument
  if(argc<2){
    cerr<<"Usage: ./HDecode Filename";
    return 0;
  }
  
  ifstream fs(argv[1]);
  if(fs.fail()) {
    cerr<<"Cannot open "<<argv[1]<<endl;
    return 0;
  }
   
  // To hide password input from screen
  //http://stackoverflow.com/questions/6899025/hide-user-input-on-password-prompt
  termios oldt;
  tcgetattr(STDIN_FILENO, &oldt);
  termios newt = oldt;
  newt.c_lflag &= ~ECHO;
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  
  while(1){
    cerr<<"Enter password:"<<endl;
    string enter;
    getline(cin, enter);
    cerr<<"Please enter again:"<<endl;
    string again;
    getline(cin, again);
    if(enter!=again){
      cerr<<"The password you entered does not match the pervious one."<<endl;
      cerr<<"Try again."<<endl;
    } else {
      setSeed(enter);	// tools.h
      break;
    }
  }

  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);	// clean up
  
  int count=0;
  char inp;
  Ascii ascii1;
  Ascii ascii2;
  vector<int> input;
  string output;
  int acc = ACCUMULATOR;

  while(1){
    ascii1=ascii2;
    input.clear();
    if(count==1) acc=ACCUMULATOR;	// need a sequence of acc acc acc*2 acc*4 acc*8 ...
    for(register int i = 0; i<acc; i++){
      fs>>noskipws>>inp;
      if(fs.fail()) break;	// works on empty input file
      input.emplace_back(inp);
      int num = (unsigned int) inp;
      ascii2.update(num);
    }
    Tree t(ascii1);
    
    if(count==0){

      // use the password as a seed to shuffle the first 128 ASCII entries.

      random_shuffle(t.begin(),t.end());	// <algorithm>
    }

    t.merge_least();	// tools.h
    output += encode(t,input);
    acc=acc*2;
    count++;
    if(fs.fail()) break;
  }
 
  // stream out binary sequence
  bin_out(output);

#ifdef DEBUG
  // tools.h
  t.huffcode();	
  ascii2.hufffreq();
  wpl(t,ascii2);
#endif

}
